src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs

```diff
@@ -136,33 +136,54 @@ protected ImmutableArray<CompletionProvider> GetProviders(ImmutableHashSet<strin
             }
         }
 
+        private ImmutableArray<CompletionProvider> GetProviders(
+            ImmutableHashSet<string> roles, CompletionTrigger trigger, OptionSet options)
+        {
+            return FilterProviders(GetProviders(roles, trigger), trigger, options);
+        }
+
         protected virtual ImmutableArray<CompletionProvider> GetProviders(
```

Comment:
Hmm - this whole trio of private/protected/protected virtual overloads seems fragile to me.  Any thoughts on how to make it obvious which ones to call?